accepted/2023/preview-apis/preview-apis.md

```diff
@@ -0,0 +1,319 @@
+to make it easier to build services. The library is distributed as a NuGet
+package. The package is already in a version well past 1.0 and is considered
+stable.
+
+He ships the package often and doesn't want to maintain multiple flavors of the
+package, one marked as prerelease and one marked as stable. However, he wants to
+expose new APIs without making the promise that these APIs are stable. He
+designs the new APIs as if they would be stable, that is, he puts them in the
+namespace they belong and names the types and members without any naming
+convention that implies that they aren't stable yet. If customer feedback is
+positive, the API will become stable as-is, otherwise he'll make changes to
+address the feedback.
+
+In order to convey to the consumers that these APIs aren't stable yet, he puts
+the `Experimental` attribute on them:
+
+```C#
+namespace MartinsLibrary.FancyLogging
+{
+    public static class FancyLoggerExtensions
+    {
+        [Experimental("ML123", UrlFormat="https://martinslibrary.net/diagnostics/{0}")]
+        public static IServiceCollection AddFancyLogging(this IServiceCollections services)
+        {
+            // ...
+        }
+    }
+}
+```
+
+### Consuming an experimental API
+
+David builds a new version of Jabber.NET and consumes Martin's library. He very
+much wants to play with the new fancy logging Martin added. When David calls
+`AddFancyLogging` he gets a compilation error:
+
+> error ML123: FancyLoggerExtensions.AddFancyLogging() is not a stable API. In
+  order to consume this experimental API, suppress this error from the call
+  site.
+
+David is happy to accommodate future breaking changes around the fancy logging
+feature. He understands that the diagnostic ID is specific to the fancy logging
+feature of Martin's library, so he decides to add a suppression to the project
+file:
+
+```xml
+<Project>
+    <PropertyGroup>
+        <!-- Fancy logging is an experimental APIs -->
+        <NoWarn>$(NoWarn);ML123</NoWarn>
+    </PropertyGroup>
+</Project>
+```
+
+### Consuming another experimental API
+
+Later, David discovers Martin's library also has support for fancy serialization.
+When he starts consuming it, he gets another compilation error:
+
+> error ML456: FancySerializationExtensions.AddFancySerialization() is not a
+  stable API. In order to consume this experimental API, suppress this error
+  from the call site.
+
+Giving it some thought, David concludes that the risk of using an unstable
+serialization API isn't worth it because it's not necessary for Jabber, so he
+stops exploring that API.
+
+## Requirements
+
+### Goals
+
+* Don't change the behavior for platform-level preview features
+
+* Offer library-level preview features with a different set of guardrails:
+    - Modelled after `[Obsolete]`, except the severity is always an error
+    - The opting-ing is expressed similar to obsolete, that is the call site is
```

Comment:
Should this have been opting-in?

Reply:
Yep, but 

<img width="455" alt="image" src="https://user-images.githubusercontent.com/5169960/225465366-2faf6b79-f554-44a5-a2e8-fccd1e163c8b.png">

Reply:
Fixed with [342c288](https://github.com/dotnet/designs/pull/290/commits/342c28871a785880bc114146809128e33d7d1bd6).